MacPass/MPPasswordInputController.m

```diff
@@ -138,6 +159,204 @@ - (IBAction)_submit:(id)sender {
   }
 }
 
+- (void) _touchIdHandleUnlockAttempt: (KPKCompositeKey*)compositeKey withResult:(bool)success {
+  if(success && self.databaseFileURL && self.databaseFileURL.lastPathComponent) {
+    NSData* encryptedKey = [self _touchIdEncryptCompositeKey:compositeKey];
+    if(encryptedKey) {
+      if (self.touchIdEnabled.state == NSControlStateValueMixed) {
+        [touchIDSecuredPasswords setObject:encryptedKey forKey:self.databaseFileURL.lastPathComponent];
+      }
+      else if(self.touchIdEnabled.state == NSControlStateValueOn) {
+        [NSUserDefaults.standardUserDefaults setObject:encryptedKey forKey:[self _userDefaultsKeyForEncryptedCompositeKey]];
+      }
+    }
+  }
+}
+
+- (void) _touchIdCreateAndAddRSAKeyPair {
+  CFErrorRef error = NULL;
+  NSString* publicKeyLabel =  @"MacPass TouchID Feature Public Key";
+  NSString* privateKeyLabel = @"MacPass TouchID Feature Private Key";
+  NSData* publicKeyTag =  [TouchIdUnlockPublicKeyTag  dataUsingEncoding:NSUTF8StringEncoding];
+  NSData* privateKeyTag = [TouchIdUnlockPrivateKeyTag dataUsingEncoding:NSUTF8StringEncoding];
+  SecAccessControlRef access = NULL;
+  if (@available(macOS 10.13.4, *)) {
+    SecAccessControlCreateFlags flags = kSecAccessControlBiometryCurrentSet;
+    if (@available(macOS 10.15, *)) {
+      flags |= kSecAccessControlWatch | kSecAccessControlOr;
+    }
+    access = SecAccessControlCreateWithFlags(kCFAllocatorDefault,
+                                             kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
+                                             flags,
+                                             &error);
+    if(access == NULL) {
+      NSError *err = CFBridgingRelease(error);
+      NSLog(@"Error while trying to create AccessControl for TouchID unlock feature: %@", [err description]);
+      return;
+    }
+    NSDictionary* attributes = @{
+      (id)kSecAttrKeyType:        (id)kSecAttrKeyTypeRSA,
+      (id)kSecAttrKeySizeInBits:  @2048,
+      (id)kSecAttrSynchronizable: @NO,
+      (id)kSecPrivateKeyAttrs:
+           @{ (id)kSecAttrIsPermanent:    @YES,
+              (id)kSecAttrApplicationTag: privateKeyTag,
+              (id)kSecAttrLabel: privateKeyLabel,
+              (id)kSecAttrAccessControl:  (__bridge id)access
+            },
+      (id)kSecPublicKeyAttrs:
+           @{ (id)kSecAttrIsPermanent:    @YES,
+              (id)kSecAttrApplicationTag: publicKeyTag,
+              (id)kSecAttrLabel: publicKeyLabel,
+            },
+    };
+    SecKeyRef privateKey = NULL;
+    SecKeyRef publicKey = NULL;
+    OSStatus result = SecKeyGeneratePair((__bridge CFDictionaryRef)attributes, &privateKey, &publicKey);
```

Comment:
Reading the headers, it is suggested to use ```SecKeyCreateRandomKey``` but this might only be usefull platforms other than macOS or do you have a specific reason to not use it?

Reply:
I looked into it and the most likely reason is, that when researching online how to create a asymmetric keypair on macOS this was the first one to turn up. I switched to the suggested function and tested it. It behaves exactly the same.